core/src/main/java/io/parsingdata/metal/token/Cho.java

```diff
@@ -43,27 +46,23 @@
 
     public Cho(final String name, final Encoding encoding, final Token token1, final Token token2, final Token... additionalTokens) {
         super(name, encoding);
-        this.tokens = create(checkContainsNoNulls(additionalTokens, "tokens"))
-                .add(checkNotNull(token2, "token2"))
-                .add(checkNotNull(token1, "token1"));
+        this.tokens = create(checkContainsNoNulls(additionalTokens, "additionalTokens"))
+            .add(checkNotNull(token2, "token2"))
+            .add(checkNotNull(token1, "token1"));
     }
 
     @Override
     protected Optional<Environment> parseImpl(final String scope, final Environment environment, final Encoding encoding) throws IOException {
-        final Optional<Environment> result = iterate(scope, environment.addBranch(this), encoding, tokens);
-        if (result.isPresent()) {
-            return success(result.get().closeBranch());
-        }
-        return failure();
+        return iterate(scope, Optional.of(environment.addBranch(this)), encoding, tokens).computeResult();
     }
 
-    private Optional<Environment> iterate(final String scope, final Environment environment, final Encoding encoding, final ImmutableList<Token> list) throws IOException {
+    private Trampoline<Optional<Environment>> iterate(final String scope, final Optional<Environment> environment, final Encoding encoding, final ImmutableList<Token> list) throws IOException {
         if (list.isEmpty()) {
-            return failure();
+            return complete(Util::failure);
         }
-        final Optional<Environment> result = list.head.parse(scope, environment, encoding);
-        if (result.isPresent()) { return result; }
-        return iterate(scope, environment, encoding, list.tail);
+        return list.head.parse(scope, environment.get(), encoding)
```

Comment:
Yes and no. There doesn't appear to be a use for an `Optional<Environment>` here at all, so I just changed it to `Environment`, which eliminates the `get()` outright.

But I did notice that there is no checking for `null` on both the `scope` and `environment` arguments (`encoding` is checked because it's explicitly allowed to be `null`). But before I make a big change like this while resolving some unrelated PR, let's investigate further: #175.